### DISCUSSION TITLE
CalcTPSA' function of including S and P seemed not working

### DISCUSSION THREAD

**Discussion by chengthefang** [rdkit/rdkit#5033] [Q&A]
Hi all,

I was calculating TPSA for a few S-contained structures. Then I turned on and off "includeSandP" flag when calling calcTPSA function, but it doesn't affect the result. Does anyone else have the same issue? See the snapshot below.

![image](https://user-images.githubusercontent.com/7895104/155244144-41bd7a83-3efd-49b9-bbc9-64830bc314c1.png)

Thanks,
Cheng

**Comment by greglandrum** [comment]
Not every S or P atom contributes to the TPSA.
Here's an example where it makes a difference:
```
In [29]: rdMolDescriptors.CalcTPSA(Chem.MolFromSmiles('CS(=O)C'),includeSandP=True)
Out[29]: 36.28

In [30]: rdMolDescriptors.CalcTPSA(Chem.MolFromSmiles('CS(=O)C'),includeSandP=False)
Out[30]: 17.07
```

**Reply by chengthefang** [reply]
> rdMolDescriptors.CalcTPSA(Chem.MolFromSmiles('CS(=O)C'),includeSandP=False)

Hi Greg,

I tried calling CalcTPSA(mol) and CalcTPSA(Chem.MolFromSmiles(smi)) with includeSandP=True on, which give me different answers. When calling CalcTPSA(mol), it seemed to remember the calculated TPSA value for the mol object. 
`
smis = ['S','SS','SSS','S=O','CSC','CS(=O)C']
mols= [Chem.MolFromSmiles(smi) for smi in smis]
for mol in mols:
    print(rdMolDescriptors.CalcTPSA(mol,includeSandP=True),rdMolDescriptors.CalcTPSA(mol,includeSandP=False))
print()
for smi in smis:
    print(rdMolDescriptors.CalcTPSA(Chem.MolFromSmiles(smi),includeSandP=True),rdMolDescriptors.CalcTPSA(Chem.MolFromSmiles(smi),includeSandP=False))`

It gave me different results:
0.0 0.0
77.6 77.6
102.89999999999999 102.89999999999999
49.160000000000004 49.160000000000004
25.3 25.3
36.28 36.28

0.0 0.0
77.6 0.0
102.89999999999999 0.0
49.160000000000004 17.07
25.3 0.0
36.28 17.07

**Reply by greglandrum** [reply]
Some of the RDKit descriptor calculators do an optimization where they cache the value of either the computed descriptor or some of the values needed to compute the descriptor on the molecule itself (this way if you calculate the descriptor a second time it goes much faster).
You can tell it to ignore that cached value with the `force` optional argument. 

Here's an example:
```
In [5]: rdMolDescriptors.CalcTPSA(mol)
Out[5]: 17.07

In [6]: rdMolDescriptors.CalcTPSA(mol,includeSandP=True)
Out[6]: 17.07

In [7]: rdMolDescriptors.CalcTPSA(mol,includeSandP=True, force=True)
Out[7]: 36.28
```

**Reply by chengthefang** [reply]
> Some of the RDKit descriptor calculators do an optimization where they cache the value of either the computed descriptor or some of the values needed to compute the descriptor on the molecule itself (this way if you calculate the descriptor a second time it goes much faster). You can tell it to ignore that cached value with the `force` optional argument.
> 
> Here's an example:
> 
> ```
> In [5]: rdMolDescriptors.CalcTPSA(mol)
> Out[5]: 17.07
> 
> In [6]: rdMolDescriptors.CalcTPSA(mol,includeSandP=True)
> Out[6]: 17.07
> 
> In [7]: rdMolDescriptors.CalcTPSA(mol,includeSandP=True, force=True)
> Out[7]: 36.28
> ```

Thanks Greg for checking that so quickly. Good to know the optimization mechanism in some rdkit calculators. It makes perfect sense.
